README.md
# <img src='https://raw.githubusercontent.com/AndyObtiva/glimmer_wordle/master/icons/linux/Glimmer Wordle.png' height=85 /> Glimmer Wordle 1.1.6
## Play Wordle Endlessly with No Limit!
[![Gem Version](https://badge.fury.io/rb/glimmer_wordle.svg)](http://badge.fury.io/rb/glimmer_wordle)

[Wordle](https://en.wikipedia.org/wiki/Wordle) word game desktop GUI app written [test-first](https://github.com/AndyObtiva/wordle/blob/master/spec/app/model/five_letter_word_spec.rb) using [Glimmer DSL for SWT](https://github.com/AndyObtiva/glimmer-dsl-swt) (JRuby Desktop Development GUI Framework) as inspiration by [Kevin Newton's blog post](https://kddnewton.com/2022/01/29/solving-wordle-in-ruby.html).

(Since the desktop version can leverage a real keyboard, there is no need to display the alphabets following the Qwerty layout, so they are displayed alphabetically to enable a more user-friendly experience, but the Qwerty layout is available as an [option](#options) if desired.)

Mac | Windows | Linux
----|---------|------
![Wordle Screenshot](screenshots/glimmer-wordle.png) | ![Wordle Screenshot Windows](screenshots/glimmer-wordle-windows.png) | ![Wordle Screenshot Linux](screenshots/glimmer-wordle-linux.png)

Other [Glimmer](https://github.com/AndyObtiva/glimmer) games:

[<img alt="Glimmer Tetris Icon" src="https://raw.githubusercontent.com/AndyObtiva/glimmer_tetris/master/package/linux/Glimmer%20Tetris.png" height=40 /> Glimmer Tetris](https://github.com/AndyObtiva/glimmer_tetris)

[<img alt="Glimmer Klondike Solitaire Icon" src="https://raw.githubusercontent.com/AndyObtiva/glimmer_klondike_solitaire/master/icons/linux/Glimmer%20Klondike%20Solitaire.png" height=40 /> Glimmer Klondike Solitaire](https://github.com/AndyObtiva/glimmer_klondike_solitaire)

## Demo

![Wordle Screenshot](screenshots/glimmer-wordle.gif)

## Setup

### Download

[<img src='https://raw.githubusercontent.com/AndyObtiva/glimmer_wordle/master/icons/linux/Glimmer Wordle.png' height=40 /> Download Glimmer Wordle 1.1.5 DMG for Mac x86_64 Ventura 13.4.1 and Older](https://www.dropbox.com/scl/fi/logokhlklr186m04lmj5u/Glimmer-Wordle-1.1.5-x86_64-ventura-13.4.1.dmg?rlkey=jlzah028gpy5h270yceeiy315&dl=1)

[<img src='https://raw.githubusercontent.com/AndyObtiva/glimmer_wordle/master/icons/linux/Glimmer Wordle.png' height=40 /> Download Glimmer Wordle 1.1.4 DMG for Mac ARM64 Ventura and Older](https://www.dropbox.com/s/diah745mycc8elz/Glimmer%20Wordle-1.1.4-arm64.dmg?dl=1)

[<img src='https://raw.githubusercontent.com/AndyObtiva/glimmer_wordle/master/icons/linux/Glimmer Wordle.png' height=40 /> Download Glimmer Wordle 1.1.2 DMG for Mac x86_64 Catalina and Older](https://www.dropbox.com/s/ysureb9o1dky7w0/Glimmer%20Wordle-1.1.2.dmg?dl=1)

[<img src='https://raw.githubusercontent.com/AndyObtiva/glimmer_wordle/master/icons/linux/Glimmer Wordle.png' height=40 /> Download Glimmer Wordle 1.1.3 DEB for Linux x86_64](https://www.dropbox.com/s/7jybodbf4pewhtd/glimmer-wordle_1.1.3-1_amd64.deb?dl=1)

[<img src='https://raw.githubusercontent.com/AndyObtiva/glimmer_wordle/master/icons/linux/Glimmer Wordle.png' height=40 /> Download Glimmer Wordle 1.1.3 RPM for Linux x86_64](https://www.dropbox.com/s/dfpjnquchxzl823/glimmer-wordle-1.1.3-1.x86_64.rpm?dl=1)

### Install Ruby Gem

If you are a software engineer and would rather install this game as a Ruby gem, please follow these instructions.

First, make sure you have all the [prerequisites of Glimmer DSL for SWT](https://github.com/AndyObtiva/glimmer-dsl-swt#pre-requisites) setup on your machine like Java and JRuby.

Run:

```
gem install glimmer_wordle
```

Afterwards, run:

```
glimmer_wordle
```

## Options

The game menu bar has these View menu options:

1- **View -> Alphabet Layout -> Alphabetical**: Displays alphabets alphabetically

![Wordle Screenshot Alphabetical](screenshots/glimmer-wordle.png)

2- **View -> Alphabet Layout -> Qwerty**: Displays alphabets with the Qwerty keyboard layout

![Wordle Screenshot Qwerty](screenshots/glimmer-wordle-qwerty.png)

Also, the game supports Dark Mode, and will adopt a Dark Mode theme automatically if you start it with your System Settings having Dark Mode.

![Wordle Mac Dark Mode](screenshots/glimmer-wordle-mac-dark-mode.png)

## Process

[Glimmer Process](https://github.com/AndyObtiva/glimmer/blob/master/PROCESS.md)

## Contributing

-   Check out the latest master to make sure the feature hasn't been
    implemented or the bug hasn't been fixed yet.
-   Check out the issue tracker to make sure someone already hasn't
    requested it and/or contributed it.
-   Fork the project.
-   Start a feature/bugfix branch.
-   Commit and push until you are happy with your contribution.
-   Make sure to add tests for it. This is important so I don't break it
    in a future version unintentionally.
-   Please try not to mess with the Rakefile, version, or history. If
    you want to have your own version, or is otherwise necessary, that
    is fine, but please isolate to its own commit so I can cherry-pick
    around it.

## TODO

[TODO.md](TODO.md)

## Change Log

[CHANGELOG.md](CHANGELOG.md)

## Copyright

[MIT](LICENSE.txt)

Copyright (c) 2022-2023 Andy Maleh. See
[LICENSE.txt](LICENSE.txt) for further details.

--

[<img src="https://raw.githubusercontent.com/AndyObtiva/glimmer/master/images/glimmer-logo-hi-res.png" height=40 />](https://github.com/AndyObtiva/glimmer) Built with [Glimmer DSL for SWT](https://github.com/AndyObtiva/glimmer-dsl-swt) (JRuby Desktop Development GUI Framework)

Glimmer Wordle icon made by <a href="https://www.flaticon.com/authors/freepik" title="Freepik">Freepik</a> from <a href="https://www.flaticon.com/" title="Flaticon">www.flaticon.com</a>


CHANGELOG.md
# Change Log

## 1.1.6

- Dark Mode support

## 1.1.5

- Tweak styling for Windows to have the app look good in it.

## 1.1.4

- Fix this issue: type one letter, navigate with right-arrow to the last letter, type another letter, hit backspace. Instead of deleting the last letter that the caret/cursor is on, it deletes the first letter instead.

## 1.1.3

- Officially support Linux with improved fonts/look and feel
- Package native executables for Linux

## 1.1.2

- Correct spelling of querty as qwerty
- Shorten help instructions to fit vertically in Monterey/Big-Sur style message box
- Ensure that the alphabet layout retains colors when switched in the middle of a game

## 1.1.1

- Fix issue with loading `clipboard` gem when running `glimmer_wordle` command from Ruby gem.

## 1.1.0

- Display answer when losing
- Close Share Emoji dialog when restarting game (with ENTER button or menu item)
- Have the main app menu bar show up in the Share Emoji dialog
- Add Help menu Instructions menu item
- Change default alphabet layout to be alphabetical (i.e. abcd...) instead of querty (since users will not be tapping like in cell phones yet using a real keyboard)
- Offer View -> Alphabet Layout menu item options of Alphabetical or Querty (storing in config file `~/.glimmer_wordle`)
- Adjust the dark yellow (to #c9b458), dark green (to #6aaa64), and dark gray (to rgb 120 124 126) colors to match the original Wordle game colors
- Support overriding a letter with typing in highlighted letter position
- Navigate letters with left and right keyboard buttons and re-enter a letter (without having to use backspace)
- Fix an issue with backspacing all the way back to the first character after only typing one character causing rectangle highlight not to move back to first letter
- Fix issue where upon restart, the guess button is showing above the entry field instead of below it
- Fix issue with backspacing making letter highlight go to last letter when word is empty

## 1.0.0

- Play wordle game
- Restart wordle game
- Share Result Emoji


spec/spec_helper.rb
# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# The generated `.rspec` file contains `--require spec_helper` which will cause
# this file to always be loaded, without a need to explicitly require it in any
# files.
#
# Given that it is always loaded, you are encouraged to keep this file as
# light-weight as possible. Requiring heavyweight dependencies from this file
# will add to the boot time of your test suite on EVERY test run, even for an
# individual file that may not need all of that loaded. Instead, consider making
# a separate helper file that requires the additional dependencies and performs
# the additional setup, and require it from the spec files that actually need
# it.
#
# The `.rspec` file also contains a few flags that are not defaults but that
# users commonly want.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
require 'bundler/setup'
Bundler.require(:default, :development)
RSpec.configure do |config|
  # The following ensures rspec tests that instantiate and set Glimmer DSL widgets in @target get cleaned after
  config.after do
    @target.dispose if @target && @target.respond_to?(:dispose)
    Glimmer::DSL::Engine.reset
  end
  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # This option will default to `true` in RSpec 4. It makes the `description`
    # and `failure_message` of custom matchers include text for helper methods
    # defined using `chain`, e.g.:
    #     be_bigger_than(2).and_smaller_than(4).description
    #     # => "be bigger than 2 and smaller than 4"
    # ...rather than:
    #     # => "be bigger than 2"
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended, and will default to
    # `true` in RSpec 4.
    mocks.verify_partial_doubles = true
  end

  # This option will default to `:apply_to_host_groups` in RSpec 4 (and will
  # have no way to turn it off -- the option exists only for backwards
  # compatibility in RSpec 3). It causes shared context metadata to be
  # inherited by the metadata hash of host groups and examples, rather than
  # triggering implicit auto-inclusion in groups with matching metadata.
  config.shared_context_metadata_behavior = :apply_to_host_groups

# The settings below are suggested to provide a good initial experience
# with RSpec, but feel free to customize to your heart's content.
=begin
  # This allows you to limit a spec run to individual examples or groups
  # you care about by tagging them with `:focus` metadata. When nothing
  # is tagged with `:focus`, all examples get run. RSpec also provides
  # aliases for `it`, `describe`, and `context` that include `:focus`
  # metadata: `fit`, `fdescribe` and `fcontext`, respectively.
  config.filter_run_when_matching :focus

  # Allows RSpec to persist some state between runs in order to support
  # the `--only-failures` and `--next-failure` CLI options. We recommend
  # you configure your source control system to ignore this file.
  config.example_status_persistence_file_path = "spec/examples.txt"

  # Limits the available syntax to the non-monkey patched syntax that is
  # recommended. For more details, see:
  #   - http://rspec.info/blog/2012/06/rspecs-new-expectation-syntax/
  #   - http://www.teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
  #   - http://rspec.info/blog/2014/05/notable-changes-in-rspec-3/#zero-monkey-patching-mode
  config.disable_monkey_patching!

  # This setting enables warnings. It's recommended, but in some cases may
  # be too noisy due to issues in dependencies.
  config.warnings = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = 'doc'
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed
=end
end

require 'glimmer/rake_task'

ROOT_PATH = File.expand_path('..', __dir__)
APP_PATH = File.join(ROOT_PATH, 'app')

require_relative '../app/wordle'


app/wordle/model/five_letter_word.rb
class Wordle
  module Model
    class FiveLetterWord
      class << self
        def random_word
          WORLD_ANSWERS_ALPHABETICAL.sample
        end
      end
      
      WORLD_ANSWERS_ALPHABETICAL_FILE = File.join(APP_ROOT, 'config', 'wordle-answers-alphabetical.txt')
      WORLD_ALLOWED_GUESSES_FILE = File.join(APP_ROOT, 'config', 'wordle-allowed-guesses.txt')
      WORLD_ANSWERS_ALPHABETICAL = File.read(WORLD_ANSWERS_ALPHABETICAL_FILE).lines(chomp: true).sort
      WORLD_ALLOWED_GUESSES = (WORLD_ANSWERS_ALPHABETICAL + File.read(WORLD_ALLOWED_GUESSES_FILE).lines(chomp: true)).sort
      COLORS = [:green, :yellow, :gray]
      STATUSES = [:in_progress, :win, :loss]
      
      attr_accessor :guesses, :guess_results, :status, :answer, :colored_alphabets
      
      def initialize
        refresh
      end
      
      def refresh
        self.guesses = []
        self.guess_results = []
        self.status = :in_progress
        self.answer = self.class.random_word
        self.colored_alphabets = (97..122).map {|n| n.chr}.reduce({}) {|hash, letter| hash.merge(letter => nil) }
      end
      
      def guess(word_guess)
        word_guess = word_guess.downcase
        return [] if status == :win || status == :loss
        guesses << word_guess
        guess_results << guess_result_for(word_guess)
        update_colored_alphabets(word_guess, guess_results.last)
        self.status = :win if guesses.last == answer
        self.status = :loss if guesses.last != answer && guesses.count == 6
        guess_results.last
      end
      
      private
      
      def guess_result_for(word_guess)
        5.times.map do |i|
          if answer[i] == word_guess[i]
            :green
          elsif answer.include?(word_guess[i])
            :yellow
          else
            :gray
          end
        end
      end
      
      def update_colored_alphabets(word_guess, guess_result)
        5.times do |i|
          if guess_result[i] == :green
            self.colored_alphabets[word_guess[i]] = :green
          elsif guess_result[i] == :yellow
            self.colored_alphabets[word_guess[i]] = :yellow if self.colored_alphabets[word_guess[i]] != :green
          elsif guess_result[i] == :gray
            self.colored_alphabets[word_guess[i]] = :gray if self.colored_alphabets[word_guess[i]].nil?
          end
        end
      end
    end
  end
end


app/wordle/view/app_view.rb
class Wordle
  module View
    class AppView
      include Glimmer::UI::CustomShell
      
      COLOR_TO_BACKGROUND_COLOR_MAP = {
        green:  rgb(106, 170, 100),
        yellow: rgb(201, 180, 88),
        gray:   rgb(120, 124, 126),
      }
      
      COLOR_TO_TEXT_COLOR_MAP = {
        green:  :white,
        yellow: :white,
        gray:   :white,
      }
      
      ALPHABET_LAYOUTS = {
        alphabetical: [
          %w[A B C D E F G H I J],
          %w[K L M N O P Q R S],
          %w[T U V W X Y Z],
        ],
        qwerty: [
          %w[Q W E R T Y U I O P],
          %w[A S D F G H J K L],
          %w[Z X C V B N M],
        ],
      }
      
      CONFIG_FILE = File.join(Dir.home, '.glimmer_wordle')
    
      before_body do
        @display = display {
          on_about {
            display_about_dialog
          }
          on_preferences {
            display_about_dialog
          }
          on_swt_keydown do |key_event|
            if key_event.keyCode == 8
              do_backspace
            elsif key_event.keyCode == swt(:arrow_left)
              do_left
            elsif key_event.keyCode == swt(:arrow_right)
              do_right
            elsif key_event.keyCode == swt(:cr)
              if @five_letter_word.status == :in_progress
                do_guess
              else
                do_restart
              end
            elsif valid_character?((key_event.keyCode.chr rescue ''))
              do_type(key_event.keyCode.chr)
            end
          end
        }
        @five_letter_word = Model::FiveLetterWord.new
        config = load_config
        @alphabet_layout = config[:alphabet_layout] || :alphabetical
      end
  
      ## Add widget content inside custom shell body
      ## Top-most widget must be a shell or another custom shell
      #
      body {
        shell(:no_resize) {
          grid_layout {
            margin_width 10
            margin_height 10
            vertical_spacing 0
          }
          
          # Replace example content below with custom shell content
          minimum_size 420, 540
          image File.join(APP_ROOT, 'icons', 'windows', "Glimmer Wordle.ico") if OS.windows?
          image File.join(APP_ROOT, 'icons', 'linux', "Glimmer Wordle.png") unless OS.windows?
          text "Glimmer Wordle"
          background Display.system_dark_theme? ? :black : :white
          
          app_menu_bar
          
          alphabet_container
          
          label {
            layout_data :center, :center, true, false
            text 'You have 6 tries to guess a 5-letter word'
            background Display.system_dark_theme? ? :black : :transparent
            background :transparent if OS.windows?
          }
          
          word_guesser
          
          guess_button
        }
      }
      
      def app_menu_bar
        menu_bar {
          menu {
            text '&Game'
            
            menu_item {
              text '&Restart'
              
              on_widget_selected {
                do_restart
              }
            }
            
            menu_item {
              text 'E&xit'
              
              on_widget_selected {
                exit(0)
              }
            }
          }
          
          menu {
            text '&View'
            
            menu {
              text 'Alphabet &Layout'
              
              menu_item(:radio) {
                text '&Alphabetical'
                selection @alphabet_layout == :alphabetical
                
                on_widget_selected {
                  self.alphabet_layout = :alphabetical
                  rebuild_alphabet_container
                }
              }
              
              menu_item(:radio) {
                text '&Qwerty'
                selection @alphabet_layout == :qwerty
                
                on_widget_selected {
                  self.alphabet_layout = :qwerty
                  rebuild_alphabet_container
                }
              }
            }
          }
          
          menu {
            text '&Help'
            
            menu_item {
              text '&Instructions'
              
              on_widget_selected {
                display_instructions_dialog
              }
            }
            
            menu_item {
              text '&About'
              
              on_widget_selected {
                display_about_dialog
              }
            }
          }
        }
      end
  
      def display_instructions_dialog
        message_box(body_root) {
          text 'Instructions'
          message <<~MULTI_LINE_STRING
            Make 6 guesses for a 5-letter word.
            
            If you enter a letter that is part of the word, and at the right location, it becomes green,
            
            If you enter a letter that is part of the word, but at the wrong location, it becomes yellow.
            
            If you enter a letter that is not part of the word, it becomes red.
          MULTI_LINE_STRING
        }.open
      end
      
      def display_about_dialog
        message_box(body_root) {
          text 'About'
          message "Glimmer Wordle #{VERSION}\n\n#{LICENSE}"
        }.open
      end
      
      def alphabet_container
        @alphabet_container = composite {
          layout_data(:center, :center, true, false)
          
          grid_layout {
            margin_width 0
            margin_height 0
            vertical_spacing 0
          }
          
          background Display.system_dark_theme? ? :black : :white
        
          alphabets
        }
      end
      
      def alphabets
        alphabet_row(ALPHABET_LAYOUTS[@alphabet_layout][0]) {
          layout_data(:center, :center, true, false) {
            width_hint 318
            height_hint 50
          }
        }
        
        alphabet_row(ALPHABET_LAYOUTS[@alphabet_layout][1]) {
          layout_data(:center, :center, true, false) {
            width_hint 288
            height_hint 50
          }
        }
        
        alphabet_row(ALPHABET_LAYOUTS[@alphabet_layout][2]) {
          layout_data(:center, :center, true, false) {
            width_hint 222
            height_hint 50
          }
        }
      end
      
      def alphabet_row(alphabet_characters, &block)
        canvas {
          block.call
          background Display.system_dark_theme? ? :black : :white
          
          @alphabet_rectangles ||= []
          @alphabet_borders ||= []
          @alphabet_letters ||= []
          alphabet_characters.each_with_index do |alphabet_character, i|
            @alphabet_rectangles << rectangle(1 + i*32, @alphabet_row_offset_y, 28, 28) {
              background Display.system_dark_theme? ? :black : :transparent
              
              @alphabet_borders << rectangle {
                foreground Display.system_dark_theme? ? :white : :gray
                line_width 2
              }
              
              @alphabet_letters << text(alphabet_character, :default, [:default, OS.linux? ? 5 : (OS.windows? ? 1 : 0)]) {
                font alphabet_font
                foreground Display.system_dark_theme? ? :white : :black
              }
            }
          end
        }
      end
      
      def alphabet_layout_alphabets
        ALPHABET_LAYOUTS[@alphabet_layout].reduce(:+)
      end
      
      def word_guesser
        @canvasses ||= []
        margin_x = 5
        margin_y = 5
        @canvasses << canvas {
          layout_data(:center, :center, true, false) {
            width_hint 230
            height_hint 50
          }
          background Display.system_dark_theme? ? :black : :white
          focus true
          
          @rectangles = []
          @borders = []
          @letters = []
          5.times do |i|
            @rectangles << rectangle(margin_x + i*45, margin_y, 40, 40) {
              background Display.system_dark_theme? ? :black : :transparent
              
              @borders << rectangle {
                foreground i == 0 ? :title_background : :gray
                line_width 2
              }
              
              @letters << text('', :default, [:default, OS.linux? ? 6 : (OS.windows? ? 1 : 0)]) {
                font letter_font
                foreground Display.system_dark_theme? ? :white : :black
              }
            }
          end
        }
      end
      
      def guess_button
        @guess_button = button {
          layout_data :center, :center, true, false
          text 'Guess'
          background Display.system_dark_theme? ? :black : :white
          foreground Display.system_dark_theme? ? :white : :black
          
          on_widget_selected do
            do_guess
          end
        }
      end
      
      def do_backspace
        @letter = @letters[highlighted_letter_index]
        if @letter.string != ''
          index_to_delete = highlighted_letter_index
        else
          index_to_delete = [highlighted_letter_index - 1, 0].max
        end
        @letter = @letters[index_to_delete]
        @letter.string = ''
        @borders.each { |caret| caret.foreground = :gray} # refactor this reusable code into a method that highlights the caret
        @borders[index_to_delete].foreground = :title_background
      end
      
      def do_guess
        return if !word_filled_up?
        word = @letters.map(&:string).join
        if invalid_word?(word)
          message_box {
            text 'Invalid Word'
            message "The word you entered is not an allowed guess!\n\nPlease try another word!"
          }.open
          return
        end
        guess_result = @five_letter_word.guess(word)
        update_guess_word_background_colors(guess_result)
        update_alphabet_background_colors
        if @five_letter_word.status == :in_progress
          @guess_button.dispose
          body_root.content {
            word_guesser
            guess_button
          }
        else
          @guess_button.dispose
          body_root.content {
            @restart_button = button {
              layout_data :center, :center, true, false
              text 'Restart'
              focus true
              
              on_widget_selected do
                do_restart
              end
            }
          }
          display_share_text_dialog
        end
        body_root.layout(true, true)
        body_root.pack(true)
      end
      
      def highlighted_letter_index
        @borders.each_with_index.find {|border, i| border.foreground.first == color(:title_background).swt_color }.last
      end
      
      def do_type(character)
        index = highlighted_letter_index
        @letter = @letters[index]
        @letter.string = character.upcase
        if @letters.any? {|letter| letter.string == ''}
          @borders.each { |caret| caret.foreground = :gray} # refactor this reusable code into a method that highlights the caret
          @borders[index == 4 ? 4 : index + 1].foreground = :title_background
        end
      end
      
      def do_left
        index = [highlighted_letter_index - 1, 0].max
        @borders.each { |caret| caret.foreground = :gray} # refactor this reusable code into a method that highlights the caret
        @borders[index].foreground = :title_background
      end
      
      def do_right
        index = [highlighted_letter_index + 1, @letters.count - 1].min
        @borders.each { |caret| caret.foreground = :gray} # refactor this reusable code into a method that highlights the caret
        @borders[index].foreground = :title_background
      end
      
      def do_restart
        @share_text_dialog&.close
        alphabet_layout_alphabets.each_with_index do |alphabet_character, i|
          @alphabet_borders[i].foreground = :gray
          @alphabet_rectangles[i].background = :white
          @alphabet_letters[i].foreground = :black
        end
        @restart_button&.dispose
        @canvasses.dup.each(&:dispose)
        @canvasses.clear
        @guess_button&.dispose
        body_root.content {
          word_guesser
          guess_button
        }
        body_root.layout(true, true)
        body_root.pack(true)
        @five_letter_word.refresh
      end
      
      def update_guess_word_background_colors(guess_result)
        guess_result.each_with_index do |result_color, i|
          background_color = COLOR_TO_BACKGROUND_COLOR_MAP[result_color]
          @borders[i].foreground = background_color
          @rectangles[i].background = background_color
          @letters[i].foreground = COLOR_TO_TEXT_COLOR_MAP[result_color]
          async_exec { @canvasses.last.redraw }
        end
      end
      
      def update_alphabet_background_colors
        alphabet_layout_alphabets.each_with_index do |alphabet_character, i|
          result_color = @five_letter_word.colored_alphabets[alphabet_character.downcase]
          if result_color
            background_color = COLOR_TO_BACKGROUND_COLOR_MAP[result_color]
            @alphabet_borders[i].foreground = background_color
            @alphabet_rectangles[i].background = background_color
            @alphabet_letters[i].foreground = COLOR_TO_TEXT_COLOR_MAP[result_color]
          end
        end
      end
      
      def word_filled_up?
        @letters.find {|letter| letter.string == ''}.nil?
      end
      
      def invalid_word?(word)
        !Model::FiveLetterWord::WORLD_ALLOWED_GUESSES.include?(word.downcase)
      end
      
      def valid_character?(character)
        ((65..90).to_a + (97..122).to_a).map {|n| n.chr}.include?(character)
      end
      
      def display_share_text_dialog
        result = "#{@five_letter_word.answer.upcase}\n\n#{emoji_result}"
          
        Clipboard.copy(result)

        @share_text_dialog = dialog(body_root) {
          grid_layout
          text 'Share Result'
          
          label {
            layout_data :center, :center, true, false
            text 'Result is copied to clipboard!'
          }
          
          styled_text {
            layout_data :fill, :fill, true, true
            editable false
            caret nil
            alignment :center
            font name: 'Segoe UI Emoji' if OS.windows?
            text result
          }
        }
        @share_text_dialog.open
      end
      
      def alphabet_font
        the_font = {style: :bold, height: 28}
        the_font.merge!(name: 'Helvetica', height: 21) if OS.linux?
        the_font.merge!(name: 'Arial', height: 25) if OS.windows?
        the_font
      end
      
      def letter_font
        the_font = {style: :bold, height: 40}
        the_font.merge!(name: 'Helvetica', height: 30) if OS.linux?
        the_font.merge!(name: 'Arial', height: 32) if OS.windows?
        the_font
      end
      
      def dispose_alphabet_container_children
        @alphabet_rectangles.clear
        @alphabet_borders.clear
        @alphabet_letters.clear
        @alphabet_container.children.each(&:dispose)
      end
      
      def rebuild_alphabet_container
        dispose_alphabet_container_children
        @alphabet_container.content {
          alphabets
        }
        @alphabet_container.layout(true, true)
        @alphabet_container.pack(true)
        update_alphabet_background_colors
      end
      
      def alphabet_layout=(value)
         @alphabet_layout = value
         save_config
      end
      
      def new_config
        {
          alphabet_layout: @alphabet_layout
        }
      end
      
      def save_config
        File.write(CONFIG_FILE, YAML.dump(new_config))
      rescue => e
        puts e.full_message
      end
      
      def load_config
        File.exist?(CONFIG_FILE) ? YAML.load(File.read(CONFIG_FILE)) : {}
      rescue => e
        puts e.full_message
        {}
      end
      
      def emoji_result
        result = ''
        @five_letter_word.guess_results.each do |row|
          row.each do |result_color|
            case result_color
            when :green
              result << "🟩"
            when :yellow
              result << "🟨"
            when :gray
              result << "⬜"
            end
          end
          result << "\n"
        end
        result
      end
    end
  end
end
